rustworkx-core/src/generators/random_graph.rs

```diff
@@ -619,14 +620,141 @@ where
     Ok(graph)
 }
 
+/// Generate a hyperbolic random undirected graph.
+///
+/// The H² hyperbolic random graph model connects pairs of nodes with a probability
+/// that decreases as their hyperbolic distance increases.
+///
+/// The number of nodes is inferred from the coordinates `radii` and `angles`. `radii`
+/// and `angles` must have the same size and cannot be empty. If `beta` is `None`,
+/// all pairs of nodes with a distance smaller than ``r`` are connected.
+///
+/// D. Krioukov et al. "Hyperbolic geometry of complex networks", Phys. Rev. E 82, pp 036106, 2010.
+///
+/// Arguments:
+///
+/// * `radii` - radial coordinates (nonnegative) of the nodes.
+/// * `angles` - angular coordinates (between -pi and pi) of the nodes.
+/// * `beta` - Sigmoid sharpness (nonnegative) of the connection probability.
+/// * `r` - Distance at which the connection probability is 0.5 for the probabilistic model.
+///     Threshold when ``beta`` is ``None``.
+/// * `seed` - An optional seed to use for the random number generator.
+/// * `default_node_weight` - A callable that will return the weight to use
+///     for newly created nodes.
+/// * `default_edge_weight` - A callable that will return the weight object
+///     to use for newly created edges.
+///
+/// # Example
+/// ```rust
+/// use rustworkx_core::petgraph;
+/// use rustworkx_core::generators::hyperbolic_random_graph;
+///
+/// let g: petgraph::graph::UnGraph<(), ()> = hyperbolic_random_graph(
+///     &vec![0.4, 2., 3.],
+///     &vec![0., 0., 0.],
+///     None,
+///     2.,
+///     None,
+///     || {()},
+///     || {()},
+/// ).unwrap();
+/// assert_eq!(g.node_count(), 3);
+/// assert_eq!(g.edge_count(), 2);
+/// ```
+pub fn hyperbolic_random_graph<G, T, F, H, M>(
+    radii: &[f64],
+    angles: &[f64],
+    beta: Option<f64>,
+    r: f64,
+    seed: Option<u64>,
+    mut default_node_weight: F,
+    mut default_edge_weight: H,
+) -> Result<G, InvalidInputError>
+where
+    G: Build + Create + Data<NodeWeight = T, EdgeWeight = M> + NodeIndexable + GraphProp,
+    F: FnMut() -> T,
+    H: FnMut() -> M,
+    G::NodeId: Eq + Hash,
+{
+    let num_nodes = radii.len();
+    if num_nodes == 0 || radii.len() != angles.len() {
+        return Err(InvalidInputError {});
+    }
+    if radii.iter().any(|x| x.is_infinite() || x.is_nan()) {
+        return Err(InvalidInputError {});
+    }
+    if angles.iter().any(|x| x.is_infinite() || x.is_nan()) {
+        return Err(InvalidInputError {});
+    }
+    if radii.iter().fold(0_f64, |a, &b| a.min(b)) < 0. {
```

Comment:
f64 doesn't implement `Ord` but you can use the workaround from https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.min to simplify